hatrac/model/directory/pgsql.py

```diff
@@ -119,15 +120,19 @@ def __setitem__(self, k, v):
         v.resource = self.resource
         v.access = k
 
-def negotiated_uri_list(uris, metadata={}):
+def negotiated_uri_list(resources, metadata={}):
     """Returns nbytes, Metadata, body"""
+    uris = [r.asurl() for r in resources]
     metadata = dict(metadata)
     metadata['content-type'] = negotiated_content_type(
-        ['application/json', 'text/uri-list'],
+        ['application/json', 'text/uri-list', 'text/html'],
         'application/json'
     )
     if metadata['content-type'] == 'text/uri-list':
         body = '\n'.join(uris) + '\n'
+    elif metadata['content-type'] == 'text/html':
+        index_of = '<h1>Index of %s</h1>\n' % os.path.dirname(resources[0].name) if resources else ''
```

Comment:
Rather than depending on a child to identify the parent, I think you should refactor this function a tiny bit further and pass in the parent resource too, since it is always known by the caller of this function.

You could add the `asurl()` method to the calling classes HatracVersions and HatracUploads so that this function could then just use something like `"Index of %s" % parent.asurl()`...